### ISSUE TITLE
Large arrays grow too much

### ISSUE THREAD

**Issue by beta-ziliani** [crystal-lang/crystal#11445]
Currently, an `Array` always [double its size](https://github.com/crystal-lang/crystal/blob/6529d725a61ed0ff83f9d9efdd18bb7229c9cebc/src/array.cr#L2028) when the capacity is exceeded. This means that when adding `@capacity+1` elements, almost half of the memory of the array is unused. This is fine for small arrays, but for larger arrays that means that, in average, at least a quarter of the array is always unused. In Go they grow the array according [to the size](https://github.com/golang/go/blob/b1b6d928bd4fb368f8ada0a554fc85405e7a3688/src/runtime/slice.go#L199-L204), from 2x to 1.25x. We should probably mimic a similar heuristic.

**Comment by HertzDevil** [comment]
Related: #6523

**Comment by beta-ziliani** [comment]
I've seen the following three growth factors:

- [MacOS `NSMutableArray`](https://ciechanow.ski/exposing-nsmutablearray/):

```swift
  newCapacity = oldCapacity * 1.625
```

This is what #6523 proposed, and there's a good explanation of why that number makes sense (in a setting where memory reuse is possible by the allocator).

- [Java's `ArrayList`](https://www.geeksforgeeks.org/internal-working-of-arraylist-in-java/):

```java
  newCapacity = oldCapacity + (oldCapacity >> 1)   // equal to `oldCapacity * 1.5`
```

Advantage: easy to calculate (no need for float conversion), and still closed to the supposedly optimal value of phi.

- [Go's `Slice`](https://github.com/golang/go/blob/b1b6d928bd4fb368f8ada0a554fc85405e7a3688/src/runtime/slice.go#L199-L204):

```go
  newcap += (newcap + 3*256) / 4
```

This is what's currently implemented in #11482.
Advantage: instead of fixing a constant, arrays grow slower as they get larger, from 2 to 1.25x. Disadvantage: it requires several more resizing than with the other two methods on large arrays.

We can start from the latter, since it's already implemented, and see how it fares. Being internal, we can change it later if we like.
